### PR TITLE
PlusCal support

### PR DESCRIPTION
**Pull request by susliko**
https://github.com/tlaplus-community/tree-sitter-tlaplus/issues/31

Block comment detection is altered from the external parser to the solution proposed in [this](https://github.com/tree-sitter/tree-sitter/discussions/1252#discussioncomment-1222142) comment. This allows for straightforward implementation of pluscal support.

@ahelwer, would be happy to hear from you

If we agree on this approach I'll proceed with docs, tests and syntax highlighting

Example: 
```tla
---- MODULE simple ----

EXTENDS TLC, Integers

(* *)   
(* comment *)      
     (* comment *)      
     (* comment *)      
     (* comment *)      
(* comment *)      
(* comment *)      
(* comment *)      
(* comment *)      


(* another comment *)
(*
just
text
(* this (* is a nested (* comment *) *)  *)

--algorithm simple
variables alice_account = 10, bob_account = 10, money = 5;
(* comment inside algorithm *)
begin
A: alice_account := alice_account - money;
B: bob_account := bob_account + money;
end algorithm; 
text
after algorithm
*)

====
```
Tree-sitter-output:
```
(source_file [0, 0] - [34, 0]
  (module [0, 0] - [32, 4]
    (header_line [0, 0] - [0, 4])
    name: (identifier [0, 12] - [0, 18])
    (header_line [0, 19] - [0, 23])
    (extends [2, 0] - [2, 21]
      (identifier_ref [2, 8] - [2, 11])
      (identifier_ref [2, 13] - [2, 21]))
    (block_comment [4, 0] - [12, 13]
      (block_comment_text [4, 2] - [12, 10]))
    (block_comment [15, 0] - [30, 2]
      (block_comment_text [15, 2] - [18, 4])
      (block_comment [19, 0] - [19, 43]
        (block_comment_text [19, 2] - [19, 7])
        (block_comment [19, 7] - [19, 39]
          (block_comment_text [19, 10] - [19, 22])
          (block_comment [19, 22] - [19, 36]
            (block_comment_text [19, 25] - [19, 33]))))
      (pcal_algorithm [21, 0] - [27, 14]
        algorithm_name: (identifier [21, 12] - [21, 18])
        (pcal_var_decls [22, 0] - [22, 58]
          (pcal_var_decl [22, 10] - [22, 29]
            (pcal_variable [22, 10] - [22, 23]
              (identifier [22, 10] - [22, 23]))
            (nat_number [22, 26] - [22, 28]))
          (pcal_var_decl [22, 30] - [22, 47]
            (pcal_variable [22, 30] - [22, 41]
              (identifier [22, 30] - [22, 41]))
            (nat_number [22, 44] - [22, 46]))
          (pcal_var_decl [22, 48] - [22, 58]
            (pcal_variable [22, 48] - [22, 53]
              (identifier [22, 48] - [22, 53]))
            (nat_number [22, 56] - [22, 57])))
        (block_comment [23, 0] - [23, 30]
          (block_comment_text [23, 2] - [23, 27]))
        (pcal_algorithm_body [24, 0] - [26, 38]
          (pcal_stmt [25, 0] - [25, 42]
            pcal_statement: (identifier [25, 0] - [25, 1])
            (pcal_unlabeled_stmt [25, 3] - [25, 42]
              (pcal_assign [25, 3] - [25, 42]
                (pcal_lhs [25, 3] - [25, 16]
                  (pcal_variable [25, 3] - [25, 16]
                    (identifier [25, 3] - [25, 16])))
                (assign [25, 17] - [25, 19])
                (bound_infix_op [25, 20] - [25, 41]
                  lhs: (identifier_ref [25, 20] - [25, 33])
                  symbol: (minus [25, 34] - [25, 35])
                  rhs: (identifier_ref [25, 36] - [25, 41])))))
          (pcal_stmt [26, 0] - [26, 38]
            pcal_statement: (identifier [26, 0] - [26, 1])
            (pcal_unlabeled_stmt [26, 3] - [26, 38]
              (pcal_assign [26, 3] - [26, 38]
                (pcal_lhs [26, 3] - [26, 14]
                  (pcal_variable [26, 3] - [26, 14]
                    (identifier [26, 3] - [26, 14])))
                (assign [26, 15] - [26, 17])
                (bound_infix_op [26, 18] - [26, 37]
                  lhs: (identifier_ref [26, 18] - [26, 29])
                  symbol: (plus [26, 30] - [26, 31])
                  rhs: (identifier_ref [26, 32] - [26, 37])))))))
      (block_comment_text [27, 14] - [29, 15]))
    (double_line [32, 0] - [32, 4])))
```

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -162,7 +174,34 @@ module.exports = grammar({
...
+    block_comment_text: $ => 
+      prec.right(repeat1(
+        choice(
+          regexOr(
+            '[^*()\n]', // any symbol except reserved
+            '[^*\n][)]', // closing parenthesis, which is not a comment end
+            '[(][^*\n]', // opening parenthesis, which is not a comment start
+            '[)][*][^)\n]',
+            '[*][^*)\n]',
+            '[*][)][ ]*[\n][(][*]' // contiguous block comment border
```
Comment (susliko) [inline grammar.js:200]: This hack is here to treat contiguous block comments as one `block_comment`

File: grammar.js [hunk 2]
```diff
@@ -162,7 +174,34 @@ module.exports = grammar({
...
+    block_comment_text: $ => 
+      prec.right(repeat1(
+        choice(
+          regexOr(
+            '[^*()\n]', // any symbol except reserved
+            '[^*\n][)]', // closing parenthesis, which is not a comment end
+            '[(][^*\n]', // opening parenthesis, which is not a comment start
+            '[)][*][^)\n]',
+            '[*][^*)\n]',
+            '[*][)][ \t]*[\n]?[ \t]*[(][*]' // contiguous block comment border
```
Comment (susliko) [inline grammar.js:200]: This hack is here to treat contiguous block comments as one block_comment

**Comment by ahelwer** [comment]
I'll take a look at this after the 25th hopefully - on holiday at the moment!

File: grammar.js (hunk 2 above)
Comment (ahelwer) [reply grammar.js:200]: This doesn't seem like it would work for CRLF line endings.

**Comment by ahelwer** [comment]
This approach looks good! Considering how complicated the `_expr` rule is it makes sense to include the pluscal grammar in this grammar.

**Comment by susliko** [comment]
@ahelwer 
I've stumbled upon 2 failing tests with conj(disj)unction lists, which are implemented via external parser.
Do you have any ideas on what should be changed to make this work?
image

**Comment by ahelwer** [comment]
I'll take a look at that failing test. This also brings to mind an edge case, which is that:
 1. Block comments can occur inside jlists
 2. PlusCal sections can occur inside block comments
 3. jlists can occur inside PlusCal sections

So I'll have to figure out a way of ensuring that the jlist in (3) doesn't interact weirdly with the jlist in (1).

**Comment by ahelwer** [comment]
I left a comment on the `scanner.cc` changes about how the lexing code for block comment tokens (not the full `scan_block_comment_text`  code, but the other stuff) needs to be left in so the non-block-comment external tokens know to bail out when they see a block comment; perhaps that will fix the failing test?

**Comment by ahelwer** [comment]
Doesn't seem like you can link to a review but try scrolling up in this PR. Here's a screenshot, I left a few other comments as well:
image

**Comment by susliko** [comment]
There is a 'Pending' comment status on the screenshot. This usually means that the review is not yet submitted (button in the upper right corner) 🤔

**Comment by ahelwer** [comment]
Well that's confusing. I think it's now in a weird state where it can't be submitted because of the force pushes overwriting commits or something.

**Review by ahelwer — CHANGES_REQUESTED** [review]
Attempting to submit review

File: src/scanner.cc [hunk 3]
```diff
@@ -136,7 +136,7 @@ namespace {
    * Checks whether the next codepoint sequence is the one given.
    * This function can change the state of the lexer.
    * 
-   * @param lexer The tree-sitter lexing control structure.
+  * @param lexer The tree-sitter lexing control structure.
```
Comment (ahelwer) [inline src/scanner.cc:139]: Accidental space deletion?

File: src/scanner.cc [hunk 4]
```diff
@@ -240,6 +240,90 @@ namespace {
...
+   *   (***********************)
+   *   (* text text text text *)
+   *   (***********************)
+   * 
+   * A block comment will also be emitted if EOF is reached.
+   *
+   * @param lexer The tree-sitter lexing control structure.
+   * @return Whether any block comment text was detected.
+   */
+  bool scan_block_comment_text(TSLexer* const lexer) {
```
Comment (ahelwer) [inline src/scanner.cc:272]: To be clear you can delete this function and the call to it. It's just the other stuff that has to stay, because the other stuff is how jlists and proofs know that they're seeing the start of a block comment.

File: src/scanner.cc [hunk 5]
```diff
@@ -1612,4 +1706,4 @@ extern "C" {
     Scanner* const scanner = static_cast<Scanner*>(payload);
     return scanner->scan(lexer, valid_symbols);
   }
-}
+}
```
Comment (ahelwer) [inline src/scanner.cc:1709]: Accidental addition of space?

**Comment by ahelwer** [comment]
Fun update - I'm working to get this grammar enabled for highlighting the TLA+ code on github! Plus fuzzy code navigation and such. Your changes will certainly be important for that.

File: queries/highlights.scm [hunk 6]
```diff
@@ -84,41 +125,51 @@
...
 (bound_nonfix_op (infix_op_symbol) @operator)
 (bound_postfix_op symbol: (_) @function.builtin)
 (bound_nonfix_op (postfix_op_symbol) @operator)
+(pcal_macro name: (identifier) @function.macro)
+(pcal_macro_call name: (identifier) @function.macro)
+(pcal_procedure name: (identifier) @function.macro)
+(pcal_algorithm name: (identifier) @namespace)
+(pcal_process name: (identifier) @function)
+(pcal_var_decl (identifier) @variable.builtin)
+("self") @variable.builtin
```
Comment (ahelwer) [inline queries/highlights.scm:158]: Unit tests picked up that this isn't a valid node type.

File: src/scanner.cc [hunk 7]
```diff
@@ -393,7 +304,6 @@ namespace {
     Lexeme_R_ANGLE_BRACKET,
     Lexeme_RIGHT_ARROW,
     Lexeme_COMMENT_START,
-    Lexeme_BLOCK_COMMENT_START,
```
[the remark this answers is not in the document]
Comment (susliko) [reply src/scanner.cc]: done

File: queries/highlights.scm [hunk 8]
```diff
@@ -84,41 +125,51 @@
...
-(comment) @comment
-(block_comment) @comment
-(single_line) @comment
-
 ; Constants, variables, and operators
 (variable_declaration (identifier) @variable.builtin)
-(constant_declaration (identifier) @constant)
-(constant_declaration (operator_declaration name: (_) @constant))
+(constant_declaration (identifier) @constant.builtin)
+(constant_declaration (operator_declaration name: (_) @constant.bulitin))
```
[the remark this answers is not in the document]
Comment (susliko) [reply queries/highlights.scm]: fixed

File: src/scanner.cc (hunk 3 above)
Comment (susliko) [reply src/scanner.cc]: yep, fixed

File: src/scanner.cc (hunk 4 above)
Comment (susliko) [reply src/scanner.cc]: deleted

File: src/scanner.cc (hunk 5 above)
Comment (susliko) [reply src/scanner.cc]: fixed

File: queries/highlights.scm (hunk 6 above)
Comment (susliko) [reply queries/highlights.scm]: fixed

File: grammar.js (hunk 2 above)
Comment (susliko) [reply grammar.js]: Not sure, if `[\r\n|\n]?` will make it work. Don't have any device running Windows to test this

Comment (ahelwer) [reply grammar.js]: You can run UTs via github actions on all platforms.

**Comment by ahelwer** [comment]
With the latest UT runs it made it all the way to the corpus tests. This is where I run the parser against every file in the https://github.com/tlaplus/examples repo to see whether it passes. Pretty easy to do locally - looks like the pluscal parsing fails on a number of them.

**Comment by susliko** [comment]
The latest update seems to parse all the specs in the corpus, except for two cases:
1. disj_list items consuming semicolon at the end of the line: 
```tla
await \/ num[nxt] = 0
      \/ <<num[self], self>> \prec <<num[nxt], nxt>> ; /* should not be consumed
```
2. Single line comments nested in block comments when the block comment ends at this same line:
```tla
(* \* *)
```

btw, I am not allowed to run UTs in this repo, [since](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) I'm a first-time contributor

File: grammar.js [hunk 9]
```diff
@@ -661,8 +707,6 @@ module.exports = grammar({
     finite_set_literal: $ => seq('{', commaList($._expr), '}'),
 
     // { x \in S : P(x) }
-    // Set dynamic precedence to 1 so that the expression {x \in S : x \in P}
```
Comment (ahelwer) [inline grammar.js:626]: Why is this comment deleted?

File: grammar.js [hunk 10]
```diff
@@ -513,7 +515,51 @@ module.exports = grammar({
...
+      $.let_in,
+      $.conj_list,
+      $.disj_list,
+    ),
+
+    /************************************************************************/
+    /* EXPRESSIONS                                                          */
+    /************************************************************************/
+
+    // Anything that evaluates to a value
```
Comment (ahelwer) [inline grammar.js:529]: What is causing this diff?

**Review by ahelwer — CHANGES_REQUESTED** [review]

**Comment by ahelwer** [comment]
(1) can probably be fixed by a simple change to the external scanner (just have it recognize a `;` character then categorize it as an `OTHER` token). (2) sounds like a puzzle!

**Comment by susliko** [comment]
> Actually for the semicolon issue, what are the desired semantics? Should the semicolon terminate the entire jlist?

Semicolon terminates a PlusCal statement, which contains the jlist. Therefore, it should terminate the jlist. As far as I understand, the lexer should emit the DEDENT token for jlist termination

**Comment by ahelwer** [comment]
Ah, okay. In that case you want to tokenize the semicolon as Token_TERMINATOR in the external scanner, which will trigger the appropriate jlist handling logic. You can probably figure out how to do this on your own but if you need help I can make the change.

**Comment by susliko** [comment]
Managed to fix both cases. All the specs in the corpus should be parsed now

**Comment by ahelwer** [comment]
Brilliant! Would you say this is ready for full review and such, or would you like to make more stylistic changes, write more tests etc.?

**Comment by susliko** [comment]
>  I see a few corpus tests are failing in the CI

Indeed. I'll have a look

> make more stylistic changes, write more tests etc.

Yeah, there are some tests to be written. I'd also like to explore the idea of getting builtin library operators (like `Head` and `Len` from the `Sequences`) highlighted only when the module `EXTENDS` this library (there is a draft in nvim's locals.scm 🤯)

**Comment by ahelwer** [comment]
That highlighting is certainly a possibility, although we can separate it from these changes.

File: grammar.js [hunk 11]
```diff
@@ -1,3 +1,9 @@
+const PREC = {
```
Comment (ahelwer) [inline grammar.js:1]: Can you use tree-sitter's partial precedence feature for this?

File: grammar.js [hunk 12]
```diff
@@ -475,7 +520,6 @@ module.exports = grammar({
       $.tuple_literal,
       $.step_expr_or_stutter,
       $.step_expr_no_stutter,
-      $.fairness,
```
Comment (ahelwer) [inline grammar.js:478]: Why is this removed?

File: grammar.js [hunk 13]
```diff
@@ -1152,5 +1196,428 @@ module.exports = grammar({
...
     ),
+
+    /**************************************************************************/
+    /* PlusCal: written inside block comments and transpiled into TLA+.       */
+    /* According to the BNF from p. 60-62 of both p-manual and c-manual       */
+    /* (https://lamport.azurewebsites.net/tla/p-manual.pdf)                   */
+    /* (https://lamport.azurewebsites.net/tla/c-manual.pdf)                   */
+    /**************************************************************************/
+
+    _pcal_algorithm: $ => choice(
```
Comment (ahelwer) [inline grammar.js:1207]: Instead of having two parallel sets of rules for the different styles, can we combine the styles in a single tree with different separators? The design philosophy of tree-sitter grammars are that they are permissive: they accept syntax that is invalid, which works fine for their use case. So if someone wants to mix separators styles for whatever reason, that's fine, we can parse that.

Comment (susliko) [reply grammar.js:1207]: I also had this thought, but the PlusCal translator enforces the choice of a single syntax(either p- or c-), giving an error on a mixed-syntax spec.
One more thing to note here is that treesitter-based refactorings (like [these](https://github.com/ThePrimeagen/refactoring.nvim)) are not syntax-agnostic. I'm not sure, but it might be useful to know whether you are having braceless syntax or not

File: grammar.js (hunk 11 above)
Comment (susliko) [reply grammar.js:1]: I tried this one. And it works when distinguishing `block_comment_text` and `pluscal_algorithm`. However it fails to distinguish `comment` and `block_comment`. Would be happy to find a way to use partial order here

File: grammar.js (hunk 13 above)
Comment (ahelwer) [reply grammar.js:1207]: Adding extra rules increases the (already large) size of the generated parser, which is a concern.

File: grammar.js (hunk 11 above)
Comment (ahelwer) [reply grammar.js:1]: If this works but partial order doesn't, then either means there's a bug in partial precedence or global order is important in the parsing of comments (since rules without precedence have default global precedence 0).

File: grammar.js (hunk 13 above)
Comment (susliko) [reply grammar.js:1207]: We might glue the two syntaxes, but it would result in correct parser outputs on invalid specs. Is this what we are ready to go for?

Comment (ahelwer) [reply grammar.js:1207]: I'm totally fine with accepting invalid syntax (we already do this with some jlist stuff). But if the size difference isn't large then sure we can keep them around. Although a secondary concern is that it makes queries less general; you have to write separate queries for C syntax and P syntax.

Comment (susliko) [reply grammar.js:1207]: I've got these numbers for the `tlaplus.so` binary
|Revision|Size (bytes) |Diff with previous (bytes)|
|---|---|---|
|TLA+ only| 3910088 | -|
|TLA+, PlusCal (P) |4761232 | 851144| 
|TLA+, PlusCal(P, C)| 4905584| 144352 |

Comment (ahelwer) [reply grammar.js:1207]: What about the generated parser.c file?

Comment (susliko) [reply grammar.js:1207]: Updated the table with parser.c sizes

File: grammar.js (hunk 12 above)
Comment (susliko) [reply grammar.js]: Fixed.
This accidental change was causing latest corpus parsing failures

File: grammar.js (hunk 11 above)
Comment (susliko) [reply grammar.js:1]: Opened a [discussion](https://github.com/tree-sitter/tree-sitter/discussions/1598)

Comment (ahelwer) [reply grammar.js:1]: Okay, we can keep it as-is until that discussion is resolved then change it later as necessary.

File: grammar.js (hunk 13 above)
Comment (ahelwer) [reply grammar.js:1208]: Been thinking about this and am leaning toward combining the trees. Pros:
- Don't need to duplicate pluscal queries; better for usability of this grammar
- Grammar file is less confusing for future developers, with fewer rules
- Avoid common problems with code duplication like fixes to one not being made to the other
- Slightly decrease size of generated C code and resulting binary (tree-sitter combines the rule trees internally)

Cons:
- We admit some invalid syntax, which isn't really a problem because we can't surface the error to the user anyway

What do you think?

File: src/scanner.cc [hunk 14]
```diff
@@ -228,108 +227,6 @@ namespace {
...
-        nest_level++;
-        GO_TO_STATE(CLexState_CONSUME);
-        END_STATE();
-      case CLexState_RIGHT_COMMENT_DELIMITER:
-        if (nest_level > 0) {
-          nest_level--;
-          GO_TO_STATE(CLexState_CONSUME);
-        } else {
-          ACCEPT_TOKEN(BLOCK_COMMENT_TEXT);
-          if (iswspace(lookahead)) GO_TO_STATE(CLexState_LOOKAHEAD);
```
Comment (ahelwer) [inline src/scanner.cc:306]: I think one or more of the macros defined at the top of the file like `GO_TO_STATE` or others were only used in this method; could you check and delete the unused macros if so?

File: grammar.js (hunk 13 above)
Comment (susliko) [reply grammar.js:1208]: I agree with the pros you've listed. However, I find it quite handy when a missing highlighting hints that  a code is not syntactically correct.

We have one more option here: aliasing (e.g. referencing both `pcal_p_if` and `pcal_c_if` as `pcal_if`).
It would address the issue of duplicate queries without giving up on accurate parsing.

Also I've already tried to reuse common rules as much as possible (e.g. [`pcal_algorithm_start`](https://github.com/susliko/tree-sitter-tlaplus/blob/6b68bde939faf9b87aeca673a9888c2ada11f2a3/grammar.js#L1244)), leaving only syntax-specific stuff like parenthesis.

One more thing to note is that merging two syntaxes is unlikely to make the grammar easier for reading, since it won't reduce the amount of code while increasing the rules complexity.

How do you like the approach of aliasing?

Comment (ahelwer) [reply grammar.js:1208]: Aliasing sounds like a good compromise!

File: src/scanner.cc (hunk 14 above)
Comment (susliko) [reply src/scanner.cc:306]: `GO_TO_STATE` is used in `scan_extramodular_text`. Others are also used

**Comment by susliko** [comment]
Although the test suite is not yet fully completed, I think the MR is ready to be reviewed 🎉

File: grammar.js [hunk 15]
```diff
@@ -1152,5 +1196,412 @@ module.exports = grammar({
...
+    ),
+
+    // fair+ process bar in 1..10
+    // variables i = 0;
+    // begin
+    // A:
+    //   i := i +1;
+    //   print(i);
+    // end process
+
```
Comment (ahelwer) [inline grammar.js:1329]: superfluous newline

File: nvim/queries/tlaplus/highlights.scm [hunk 16]
```diff
@@ -112,44 +193,13 @@
...
+  "."
   "!"
+  ";"
+  (bullet_conj)
+  (bullet_disj)
 ] @punctuation.delimiter
-(bullet_conj) @punctuation.special
-(bullet_disj) @punctuation.special
-(prev_func_val) @punctuation.special
-(placeholder) @punctuation.special
```
Comment (ahelwer) [inline nvim/queries/tlaplus/highlights.scm:122]: Should these two tokens have any defined highlighting?

**Comment by ahelwer** [comment]
Looks great! Can't wait to see reference-based highlighting working. @hwayne should be interested to hear of this progress.

File: nvim/queries/tlaplus/highlights.scm (hunk 16 above)
Comment (susliko) [reply nvim/queries/tlaplus/highlights.scm:122]: fixed

File: grammar.js (hunk 15 above)
Comment (susliko) [reply grammar.js]: fixed

**Comment by susliko** [comment]
The test suite is completed now

File: test/corpus/pluscal/assert.txt [hunk 17]
```diff
@@ -0,0 +1,25 @@
+================================================================================
```
Comment (ahelwer) [inline test/corpus/pluscal/assert.txt:1]: I'm surprised these tests actually work; this isn't yet a documented tree-sitter feature, but using a suffix to the `===` and `---` like `|||` ensures you can disambiguate the TLA+ syntax from the tree-sitter test file syntax (this feature was actually added after encountering this problem).

Comment (susliko) [reply test/corpus/pluscal/assert.txt:1]: Yeah, they both run all the tests and detect errors

Comment (ahelwer) [reply test/corpus/pluscal/assert.txt:1]: All right, just for robustness reasons I think we should add the suffixes though.

Comment (susliko) [reply test/corpus/pluscal/assert.txt]: Fixed

File: test/corpus/pluscal/assert.txt [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+PlusCal Assert
+==================|||
+---- MODULE Test ----
+(* --algorithm Test
+begin
+  assert TRUE
+end algorithm *)
+====
+
+--------------------------------------------------------------------------------
```
Comment (ahelwer) [inline test/corpus/pluscal/assert.txt:11]: Need to add the suffix on these separators as well

Comment (susliko) [reply test/corpus/pluscal/assert.txt]: done

**Review by ahelwer — APPROVED** [review]

**Comment by ahelwer** [comment]
Great! Feel free to announce this release [on the google group thread](https://groups.google.com/g/tlaplus/c/WJMVJN55GGI).

You can also issue a PR against the nvim-treesitter repo and tag me.

**Comment by susliko** [comment]
@ahelwer Thank you for assisting throughout this work!

**Comment by ahelwer** [comment]
@susliko thanks for all the hard work! I'm sure many people will appreciate support for pluscal.

**Comment by ahelwer** [comment]
@susliko could you please backport the changes to the neovim query files to this repo? Thanks!

**Comment by susliko** [comment]
Yeah, sure
